# RDD creation

#### [Introduction to Spark with Python, by Jose A. Dianes](https://github.com/jadianes/spark-py-notebooks)

Apache Spark trabaja con un conjunto de datos denominados RDD (Resilient Distributed Dataset o Conjunto de Datos Distribuidos Resistentes), estos poseen una serie de características que los hacen diferenciarse de otros tipos de estructuras de datos:
  + Inmutables: Una vez creados no se pueden modificar.
  + Distribuidos: Hace referencia al RDD, están divididos en particiones que están repartidas por el clúster
  + Resilientes: Esto quiere decir que en el caso de que se pierda una partición, esta se regenara automáticamente.

Los RDD a pesar de ser inmutables pueden ser transformados, de manera que se crean un nuevo RDD y estas transformaciones se aplican a los datos del nuevo RDD.

Existen distintas formas de generar RDDs:
  + A partir de un fichero
  + Distribución de datos desde el driver
  + Transformar un RDD para crear un nuevo RDD.

## Ciclo de vida de un RDD

![ciclo de vida de RDD](https://keepcoding.io/wp-content/uploads/2022/06/image-39-1024x473.png)

# SparkContext

SparkContext o Punto de acceso. 

Para realizar operaciones necesitamos un Context: 
  + SparkContext, SQLContext...

Dependerá del tipo de operación al principio estaba SparkContext y se usaba para operaciones con RDDs, despues salio SparkSession, para RDDs, Dataframes y Datsets. 

SparkSession contempla internamente el SparkContext, HiveContext, SQLContext...

SparkSession nos sirve para todos los contextos.

En principio usar SparkSession sería lo más correcto, ya que establecemos una sesión con el nodo maestro.

In [0]:

import pyspark

sc = spark.Context

sc

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<command-497020945842112> in <module>
      1 import pyspark
      2 
----> 3 sc = spark.Context
      4 
      5 sc

AttributeError: 'SparkSession' object has no attribute 'Context'

In [0]:
spark = SparkSession.builder.appName('SparkByExamples.com').getOrCreate()

In [0]:
sc = spark.sparkContext

In [0]:
sc

SparkContext 

 Spark UI 

 
 Version 
 v3.2.1 
 Master 
 local[8] 
 AppName 
 Databricks Shell

En SparkContext disponemos de una interfaz gráfica donde se encuentran todos los jobs ejecutados

# Obtención de los datos

Vamos a usar una parte (10%) del dataset utilizado en la KDD Cup de 1999, contiene aproximadamente medio millón de filas. La vamos a descargar de forma local en un archivo Gzip

### Crear un RDD desde un fichero

In [0]:
url = "http://kdd.ics.uci.edu/databases/kddcup99/kddcup.data_10_percent.gz"
from pyspark import SparkFiles
spark.sparkContext.addFile(url)



### Crear un RDD usando paralelización

In [0]:
myRDD = sc.textFile("file://"+SparkFiles.get('kddcup.data_10_percent.gz'))


In [0]:
myRDD.take(5)

Out[11]: ['0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,239,486,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,19,19,1.00,0.00,0.05,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,235,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,29,29,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,219,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,39,39,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,217,2032,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,49,49,1.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,normal.']

In [0]:
type(myRDD)

Out[12]: pyspark.rdd.RDD

In [0]:
myRDD.count()

Out[13]: 494021

###Obtener datos y particiones

In [0]:
a = range(100)

data = sc.parallelize(a,8)
type(data)


Out[14]: pyspark.rdd.PipelinedRDD

In [0]:
rddCollet = data.collect()
print("Number of Partitions: "+str(data.getNumPartitions()))
print("Action: First element: "+str(data.first()))
print(rddCollet)

Number of Partitions: 8
Action: First element: 0
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]


# Transformaciones de datos

### Filter

Con esta tipo de transformación podemos generar un RDD con los elementos que cumplan una condición determinada

In [0]:
myRDD.count()


Out[19]: 494021

In [0]:
normal_myRDD = myRDD.filter(lambda x : 'normal' in x)

In [0]:
normal_myRDD.count()

Out[21]: 97278

In [0]:
from time import time
t0 = time()
normal_count = normal_myRDD.count()
tt = time() - t0
print("Hay {} interaciones etiquetadas como 'normal'".format(normal_count))
print("Count completedo en {} segundos".format(round(tt,3)))

Hay 97278 interaciones etiquetadas como 'normal'
Count completedo en 1.753 segundos


Como vemos el número total de filas ha cambiado de nuestro dataset original que tenía 494021 filas a 97278.

Como norma general los calculos se realizan cuando ejecutamos una acción, y no cuando realizamos transformaciones

### Map

Al igual que en Python la función Map nos permite aplicar una función a cada elemento de nuestro RDD.
Las expresiones lambda son muy útiles para este tipo de transformaciones.

In [0]:
csv_data = myRDD.map(lambda x: x.split(","))

t0 = time()
head_rows = csv_data.take(5)
tt = time() - t0
print("Completedo en {} segundos".format(round(tt,3)))



Completedo en 0.865 segundos


In [0]:
def parse_interaction(line):
    elems = line.split(",")
    tag = elems[41]
    return (tag, elems)

key_csv_data = myRDD.map(parse_interaction)
head_rows = key_csv_data.take(5)
print(head_rows[0:2])

[('normal.', ['0', 'tcp', 'http', 'SF', '181', '5450', '0', '0', '0', '0', '0', '1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '8', '8', '0.00', '0.00', '0.00', '0.00', '1.00', '0.00', '0.00', '9', '9', '1.00', '0.00', '0.11', '0.00', '0.00', '0.00', '0.00', '0.00', 'normal.']), ('normal.', ['0', 'tcp', 'http', 'SF', '239', '486', '0', '0', '0', '0', '0', '1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '8', '8', '0.00', '0.00', '0.00', '0.00', '1.00', '0.00', '0.00', '19', '19', '1.00', '0.00', '0.05', '0.00', '0.00', '0.00', '0.00', '0.00', 'normal.'])]


### The Collect action

Collect es otro método de los báscicos que debemos conocer, básicamente lo que hace es coger todos los elementos de un RDD y los carga en nuestra memoria, para poder trabajar con ellos de forma local. 

Debemos ser precavidos cuando lo usemos, especialmente cuando estemos trabajando con RDDs con un gran número de filas.

In [0]:

t0 = time()
all_raw_data = myRDD.collect()
tt = time() - t0
print("Data collected in {} seconds".format(round(tt,3)))


Data collected in 4.305 seconds


In [0]:
all_raw_data[:3]


Out[26]: ['0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,239,486,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,19,19,1.00,0.00,0.05,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,235,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,29,29,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.']

# Conceptos Básicos de Spark DataFrame

Vamos a cargar unos datos desde la base de datos spark

In [0]:
df = spark.read.json('dbfs:/databricks-datasets/structured-streaming/events/file-0.json')


In [0]:
type(df)


Out[28]: pyspark.sql.dataframe.DataFrame

### Mostrar los datos

In [0]:
df.show()


+------+----------+
|action|      time|
+------+----------+
|  Open|1469501107|
|  Open|1469501147|
|  Open|1469501202|
|  Open|1469501219|
|  Open|1469501225|
|  Open|1469501234|
|  Open|1469501245|
|  Open|1469501246|
|  Open|1469501248|
|  Open|1469501256|
|  Open|1469501264|
|  Open|1469501266|
|  Open|1469501267|
|  Open|1469501269|
|  Open|1469501271|
|  Open|1469501282|
|  Open|1469501285|
|  Open|1469501291|
|  Open|1469501297|
|  Open|1469501303|
+------+----------+
only showing top 20 rows



In [0]:
display(df)

action time Open 1469501107 Open 1469501147 Open 1469501202 Open 1469501219 Open 1469501225 Open 1469501234 Open 1469501245 Open 1469501246 Open 1469501248 Open 1469501256 Open 1469501264 Open 1469501266 Open 1469501267 Open 1469501269 Open 1469501271 Open 1469501282 Open 1469501285 Open 1469501291 Open 1469501297 Open 1469501303 Open 1469501322 Open 1469501335 Open 1469501344 Open 1469501346 Open 1469501349 Open 1469501357 Open 1469501366 Open 1469501371 Open 1469501375 Open 1469501375 Open 1469501381 Open 1469501392 Open 1469501402 Open 1469501407 Open 1469501410 Open 1469501420 Open 1469501424 Open 1469501438 Close 1469501442 Open 1469501462 Open 1469501480 Open 1469501488 Open 1469501489 Open 1469501491 Open 1469501503 Open 1469501505 Open 1469501509 Open 1469501513 Open 1469501517 Open 1469501520 Open 1469501525 Open 1469501533 Open 1469501539 Open 1469501540 Open 1469501541 Open 1469501543 Open 1469501544 Close 1469501545 Open 1469501545 Open 1469501547 Open 1469501552 Open 1469501557 Open 1469501559 Open 1469501560 Open 1469501560 Open 1469501565 Open 1469501566 Open 1469501574 Open 1469501575 Open 1469501575 Open 1469501578 Open 1469501581 Open 1469501584 Open 1469501600 Open 1469501601 Open 1469501603 Open 1469501610 Open 1469501620 Open 1469501621 Open 1469501625 Close 1469501625 Open 1469501626 Open 1469501631 Open 1469501632 Open 1469501632 Open 1469501638 Open 1469501643 Open 1469501646 Open 1469501662 Open 1469501662 Open 1469501662 Open 1469501663 Open 1469501667 Open 1469501674 Open 1469501675 Close 1469501678 Open 1469501680 Open 1469501685 Open 1469501686 Open 1469501689 Open 1469501691 Open 1469501694 Close 1469501696 Open 1469501702 Open 1469501703 Open 1469501704 Open 1469501706 Open 1469501706 Open 1469501710 Open 1469501715 Open 1469501717 Open 1469501719 Open 1469501719 Open 1469501734 Open 1469501739 Open 1469501740 Open 1469501747 Open 1469501749 Close 1469501749 Open 1469501754 Open 1469501755 Open 1469501756 Open 1469501756 Open 1469501757 Open 1469501758 Open 1469501759 Open 1469501761 Open 1469501764 Open 1469501772 Open 1469501772 Close 1469501776 Open 1469501780 Open 1469501782 Open 1469501783 Open 1469501785 Open 1469501789 Open 1469501795 Open 1469501802 Open 1469501802 Open 1469501806 Open 1469501813 Open 1469501817 Open 1469501818 Close 1469501819 Open 1469501828 Open 1469501829 Open 1469501830 Open 1469501833 Open 1469501835 Open 1469501837 Open 1469501838 Open 1469501840 Open 1469501845 Open 1469501848 Open 1469501853 Open 1469501855 Close 1469501861 Open 1469501861 Open 1469501862 Open 1469501863 Open 1469501865 Open 1469501873 Open 1469501884 Open 1469501895 Open 1469501904 Open 1469501907 Close 1469501909 Open 1469501909 Open 1469501911 Open 1469501929 Open 1469501930 Open 1469501930 Open 1469501931 Open 1469501935 Open 1469501935 Open 1469501946 Open 1469501946 Open 1469501959 Open 1469501967 Close 1469501972 Open 1469501976 Open 1469501978 Open 1469501978 Open 1469501978 Open 1469501980 Open 1469501980 Open 1469501985 Open 1469501988 Open 1469501992 Open 1469501996 Open 1469502005 Open 1469502010 Close 1469502014 Open 1469502020 Open 1469502022 Open 1469502022 Open 1469502031 Open 1469502031 Open 1469502033 Open 1469502035 Open 1469502038 Open 1469502044 Open 1469502054 Open 1469502054 Open 1469502054 Open 1469502057 Open 1469502060 Open 1469502065 Open 1469502067 Open 1469502071 Open 1469502071 Close 1469502072 Open 1469502073 Open 1469502077 Open 1469502080 Open 1469502092 Open 1469502097 Open 1469502105 Open 1469502109 Open 1469502118 Open 1469502126 Open 1469502127 Open 1469502130 Open 1469502130 Open 1469502132 Open 1469502135 Open 1469502144 Open 1469502145 Open 1469502147 Close 1469502148 Open 1469502154 Open 1469502157 Open 1469502165 Open 1469502177 Open 1469502181 Open 1469502181 Open 1469502182 Open 1469502184 Open 1469502184 Open 1469502190 Open 1469502194 Open 1469502201 Open 1469502202 Open 1469502205 Open 1469502206 Open 1469502211 Open 1469502217 Open 1469502218 Open 1

In [0]:
df.printSchema()

root
 |-- action: string (nullable = true)
 |-- time: long (nullable = true)



In [0]:
df.columns

Out[32]: ['action', 'time']

In [0]:
df.describe()

Out[34]: DataFrame[summary: string, action: string, time: string]

### Operaciones con columnas

#### Crear una columna

In [0]:
df['time']


Out[35]: Column<'time'>

In [0]:
type(df['time'])

Out[36]: pyspark.sql.column.Column

In [0]:
df.select('time').show()

+----------+
|      time|
+----------+
|1469501107|
|1469501147|
|1469501202|
|1469501219|
|1469501225|
|1469501234|
|1469501245|
|1469501246|
|1469501248|
|1469501256|
|1469501264|
|1469501266|
|1469501267|
|1469501269|
|1469501271|
|1469501282|
|1469501285|
|1469501291|
|1469501297|
|1469501303|
+----------+
only showing top 20 rows



In [0]:
display(df.select('time'))

time 1469501107 1469501147 1469501202 1469501219 1469501225 1469501234 1469501245 1469501246 1469501248 1469501256 1469501264 1469501266 1469501267 1469501269 1469501271 1469501282 1469501285 1469501291 1469501297 1469501303 1469501322 1469501335 1469501344 1469501346 1469501349 1469501357 1469501366 1469501371 1469501375 1469501375 1469501381 1469501392 1469501402 1469501407 1469501410 1469501420 1469501424 1469501438 1469501442 1469501462 1469501480 1469501488 1469501489 1469501491 1469501503 1469501505 1469501509 1469501513 1469501517 1469501520 1469501525 1469501533 1469501539 1469501540 1469501541 1469501543 1469501544 1469501545 1469501545 1469501547 1469501552 1469501557 1469501559 1469501560 1469501560 1469501565 1469501566 1469501574 1469501575 1469501575 1469501578 1469501581 1469501584 1469501600 1469501601 1469501603 1469501610 1469501620 1469501621 1469501625 1469501625 1469501626 1469501631 1469501632 1469501632 1469501638 1469501643 1469501646 1469501662 1469501662 1469501662 1469501663 1469501667 1469501674 1469501675 1469501678 1469501680 1469501685 1469501686 1469501689 1469501691 1469501694 1469501696 1469501702 1469501703 1469501704 1469501706 1469501706 1469501710 1469501715 1469501717 1469501719 1469501719 1469501734 1469501739 1469501740 1469501747 1469501749 1469501749 1469501754 1469501755 1469501756 1469501756 1469501757 1469501758 1469501759 1469501761 1469501764 1469501772 1469501772 1469501776 1469501780 1469501782 1469501783 1469501785 1469501789 1469501795 1469501802 1469501802 1469501806 1469501813 1469501817 1469501818 1469501819 1469501828 1469501829 1469501830 1469501833 1469501835 1469501837 1469501838 1469501840 1469501845 1469501848 1469501853 1469501855 1469501861 1469501861 1469501862 1469501863 1469501865 1469501873 1469501884 1469501895 1469501904 1469501907 1469501909 1469501909 1469501911 1469501929 1469501930 1469501930 1469501931 1469501935 1469501935 1469501946 1469501946 1469501959 1469501967 1469501972 1469501976 1469501978 1469501978 1469501978 1469501980 1469501980 1469501985 1469501988 1469501992 1469501996 1469502005 1469502010 1469502014 1469502020 1469502022 1469502022 1469502031 1469502031 1469502033 1469502035 1469502038 1469502044 1469502054 1469502054 1469502054 1469502057 1469502060 1469502065 1469502067 1469502071 1469502071 1469502072 1469502073 1469502077 1469502080 1469502092 1469502097 1469502105 1469502109 1469502118 1469502126 1469502127 1469502130 1469502130 1469502132 1469502135 1469502144 1469502145 1469502147 1469502148 1469502154 1469502157 1469502165 1469502177 1469502181 1469502181 1469502182 1469502184 1469502184 1469502190 1469502194 1469502201 1469502202 1469502205 1469502206 1469502211 1469502217 1469502218 1469502229 1469502231 1469502231 1469502234 1469502236 1469502241 1469502244 1469502245 1469502246 1469502253 1469502257 1469502258 1469502259 1469502259 1469502261 1469502267 1469502269 1469502269 1469502270 1469502272 1469502272 1469502273 1469502273 1469502275 1469502277 1469502279 1469502279 1469502282 1469502285 1469502286 1469502292 1469502294 1469502298 1469502301 1469502302 1469502304 1469502308 1469502318 1469502323 1469502328 1469502333 1469502336 1469502338 1469502346 1469502348 1469502350 1469502351 1469502357 1469502361 1469502361 1469502364 1469502365 1469502367 1469502369 1469502372 1469502374 1469502377 1469502379 1469502379 1469502382 1469502385 1469502388 1469502404 1469502411 1469502416 1469502416 1469502417 1469502422 1469502429 1469502430 1469502430 1469502432 1469502432 1469502433 1469502444 1469502445 1469502446 1469502446 1469502453 1469502456 1469502464 1469502470 1469502471 1469502472 1469502474 1469502475 1469502480 1469502481 1469502490 1469502497 1469502497 1469502497 1469502500 1469502500 1469502501 1469502507 1469502507 1469502508 1469502512 1469502514 1469502515 1469502517 1469502527 1469502527 1469502529 1469502538 1469502549 1469502553 1469502555 1469502560 1469502561 1469502561 1469502562 1469502564 1469502573 14

In [0]:
df.head(2)

Out[39]: [Row(action='Open', time=1469501107), Row(action='Open', time=1469501147)]

#### Renombrar una columna

In [0]:
display(df.withColumnRenamed('action', 'superaction'))


superaction time Open 1469501107 Open 1469501147 Open 1469501202 Open 1469501219 Open 1469501225 Open 1469501234 Open 1469501245 Open 1469501246 Open 1469501248 Open 1469501256 Open 1469501264 Open 1469501266 Open 1469501267 Open 1469501269 Open 1469501271 Open 1469501282 Open 1469501285 Open 1469501291 Open 1469501297 Open 1469501303 Open 1469501322 Open 1469501335 Open 1469501344 Open 1469501346 Open 1469501349 Open 1469501357 Open 1469501366 Open 1469501371 Open 1469501375 Open 1469501375 Open 1469501381 Open 1469501392 Open 1469501402 Open 1469501407 Open 1469501410 Open 1469501420 Open 1469501424 Open 1469501438 Close 1469501442 Open 1469501462 Open 1469501480 Open 1469501488 Open 1469501489 Open 1469501491 Open 1469501503 Open 1469501505 Open 1469501509 Open 1469501513 Open 1469501517 Open 1469501520 Open 1469501525 Open 1469501533 Open 1469501539 Open 1469501540 Open 1469501541 Open 1469501543 Open 1469501544 Close 1469501545 Open 1469501545 Open 1469501547 Open 1469501552 Open 1469501557 Open 1469501559 Open 1469501560 Open 1469501560 Open 1469501565 Open 1469501566 Open 1469501574 Open 1469501575 Open 1469501575 Open 1469501578 Open 1469501581 Open 1469501584 Open 1469501600 Open 1469501601 Open 1469501603 Open 1469501610 Open 1469501620 Open 1469501621 Open 1469501625 Close 1469501625 Open 1469501626 Open 1469501631 Open 1469501632 Open 1469501632 Open 1469501638 Open 1469501643 Open 1469501646 Open 1469501662 Open 1469501662 Open 1469501662 Open 1469501663 Open 1469501667 Open 1469501674 Open 1469501675 Close 1469501678 Open 1469501680 Open 1469501685 Open 1469501686 Open 1469501689 Open 1469501691 Open 1469501694 Close 1469501696 Open 1469501702 Open 1469501703 Open 1469501704 Open 1469501706 Open 1469501706 Open 1469501710 Open 1469501715 Open 1469501717 Open 1469501719 Open 1469501719 Open 1469501734 Open 1469501739 Open 1469501740 Open 1469501747 Open 1469501749 Close 1469501749 Open 1469501754 Open 1469501755 Open 1469501756 Open 1469501756 Open 1469501757 Open 1469501758 Open 1469501759 Open 1469501761 Open 1469501764 Open 1469501772 Open 1469501772 Close 1469501776 Open 1469501780 Open 1469501782 Open 1469501783 Open 1469501785 Open 1469501789 Open 1469501795 Open 1469501802 Open 1469501802 Open 1469501806 Open 1469501813 Open 1469501817 Open 1469501818 Close 1469501819 Open 1469501828 Open 1469501829 Open 1469501830 Open 1469501833 Open 1469501835 Open 1469501837 Open 1469501838 Open 1469501840 Open 1469501845 Open 1469501848 Open 1469501853 Open 1469501855 Close 1469501861 Open 1469501861 Open 1469501862 Open 1469501863 Open 1469501865 Open 1469501873 Open 1469501884 Open 1469501895 Open 1469501904 Open 1469501907 Close 1469501909 Open 1469501909 Open 1469501911 Open 1469501929 Open 1469501930 Open 1469501930 Open 1469501931 Open 1469501935 Open 1469501935 Open 1469501946 Open 1469501946 Open 1469501959 Open 1469501967 Close 1469501972 Open 1469501976 Open 1469501978 Open 1469501978 Open 1469501978 Open 1469501980 Open 1469501980 Open 1469501985 Open 1469501988 Open 1469501992 Open 1469501996 Open 1469502005 Open 1469502010 Close 1469502014 Open 1469502020 Open 1469502022 Open 1469502022 Open 1469502031 Open 1469502031 Open 1469502033 Open 1469502035 Open 1469502038 Open 1469502044 Open 1469502054 Open 1469502054 Open 1469502054 Open 1469502057 Open 1469502060 Open 1469502065 Open 1469502067 Open 1469502071 Open 1469502071 Close 1469502072 Open 1469502073 Open 1469502077 Open 1469502080 Open 1469502092 Open 1469502097 Open 1469502105 Open 1469502109 Open 1469502118 Open 1469502126 Open 1469502127 Open 1469502130 Open 1469502130 Open 1469502132 Open 1469502135 Open 1469502144 Open 1469502145 Open 1469502147 Close 1469502148 Open 1469502154 Open 1469502157 Open 1469502165 Open 1469502177 Open 1469502181 Open 1469502181 Open 1469502182 Open 1469502184 Open 1469502184 Open 1469502190 Open 1469502194 Open 1469502201 Open 1469502202 Open 1469502205 Open 1469502206 Open 1469502211 Open 1469502217 Open 1469502218 O

#### Operaciones con columnas

In [0]:
display(df.withColumn('newtime', df['time']*2))

action time newtime Open 1469501107 2939002214 Open 1469501147 2939002294 Open 1469501202 2939002404 Open 1469501219 2939002438 Open 1469501225 2939002450 Open 1469501234 2939002468 Open 1469501245 2939002490 Open 1469501246 2939002492 Open 1469501248 2939002496 Open 1469501256 2939002512 Open 1469501264 2939002528 Open 1469501266 2939002532 Open 1469501267 2939002534 Open 1469501269 2939002538 Open 1469501271 2939002542 Open 1469501282 2939002564 Open 1469501285 2939002570 Open 1469501291 2939002582 Open 1469501297 2939002594 Open 1469501303 2939002606 Open 1469501322 2939002644 Open 1469501335 2939002670 Open 1469501344 2939002688 Open 1469501346 2939002692 Open 1469501349 2939002698 Open 1469501357 2939002714 Open 1469501366 2939002732 Open 1469501371 2939002742 Open 1469501375 2939002750 Open 1469501375 2939002750 Open 1469501381 2939002762 Open 1469501392 2939002784 Open 1469501402 2939002804 Open 1469501407 2939002814 Open 1469501410 2939002820 Open 1469501420 2939002840 Open 1469501424 2939002848 Open 1469501438 2939002876 Close 1469501442 2939002884 Open 1469501462 2939002924 Open 1469501480 2939002960 Open 1469501488 2939002976 Open 1469501489 2939002978 Open 1469501491 2939002982 Open 1469501503 2939003006 Open 1469501505 2939003010 Open 1469501509 2939003018 Open 1469501513 2939003026 Open 1469501517 2939003034 Open 1469501520 2939003040 Open 1469501525 2939003050 Open 1469501533 2939003066 Open 1469501539 2939003078 Open 1469501540 2939003080 Open 1469501541 2939003082 Open 1469501543 2939003086 Open 1469501544 2939003088 Close 1469501545 2939003090 Open 1469501545 2939003090 Open 1469501547 2939003094 Open 1469501552 2939003104 Open 1469501557 2939003114 Open 1469501559 2939003118 Open 1469501560 2939003120 Open 1469501560 2939003120 Open 1469501565 2939003130 Open 1469501566 2939003132 Open 1469501574 2939003148 Open 1469501575 2939003150 Open 1469501575 2939003150 Open 1469501578 2939003156 Open 1469501581 2939003162 Open 1469501584 2939003168 Open 1469501600 2939003200 Open 1469501601 2939003202 Open 1469501603 2939003206 Open 1469501610 2939003220 Open 1469501620 2939003240 Open 1469501621 2939003242 Open 1469501625 2939003250 Close 1469501625 2939003250 Open 1469501626 2939003252 Open 1469501631 2939003262 Open 1469501632 2939003264 Open 1469501632 2939003264 Open 1469501638 2939003276 Open 1469501643 2939003286 Open 1469501646 2939003292 Open 1469501662 2939003324 Open 1469501662 2939003324 Open 1469501662 2939003324 Open 1469501663 2939003326 Open 1469501667 2939003334 Open 1469501674 2939003348 Open 1469501675 2939003350 Close 1469501678 2939003356 Open 1469501680 2939003360 Open 1469501685 2939003370 Open 1469501686 2939003372 Open 1469501689 2939003378 Open 1469501691 2939003382 Open 1469501694 2939003388 Close 1469501696 2939003392 Open 1469501702 2939003404 Open 1469501703 2939003406 Open 1469501704 2939003408 Open 1469501706 2939003412 Open 1469501706 2939003412 Open 1469501710 2939003420 Open 1469501715 2939003430 Open 1469501717 2939003434 Open 1469501719 2939003438 Open 1469501719 2939003438 Open 1469501734 2939003468 Open 1469501739 2939003478 Open 1469501740 2939003480 Open 1469501747 2939003494 Open 1469501749 2939003498 Close 1469501749 2939003498 Open 1469501754 2939003508 Open 1469501755 2939003510 Open 1469501756 2939003512 Open 1469501756 2939003512 Open 1469501757 2939003514 Open 1469501758 2939003516 Open 1469501759 2939003518 Open 1469501761 2939003522 Open 1469501764 2939003528 Open 1469501772 2939003544 Open 1469501772 2939003544 Close 1469501776 2939003552 Open 1469501780 2939003560 Open 1469501782 2939003564 Open 1469501783 2939003566 Open 1469501785 2939003570 Open 1469501789 2939003578 Open 1469501795 2939003590 Open 1469501802 2939003604 Open 1469501802 2939003604 Open 1469501806 2939003612 Open 1469501813 2939003626 Open 1469501817 2939003634 Open 1469501818 2939003636 Close 1469501819 2939003638 Open 1469501828 2939003656 Open 1469501829 2939003658 Open 1469501830 2939003660 Ope

## SQL

Podemos usar queries de SQL directamente sobre el dataframe, pero para ellos deberemos de guardarlo en una vista temporal.

In [0]:
df.createOrReplaceTempView('IoT')

In [0]:
sql_results = spark.sql('SELECT * FROM IoT')

In [0]:
display(sql_results)

action time Open 1469501107 Open 1469501147 Open 1469501202 Open 1469501219 Open 1469501225 Open 1469501234 Open 1469501245 Open 1469501246 Open 1469501248 Open 1469501256 Open 1469501264 Open 1469501266 Open 1469501267 Open 1469501269 Open 1469501271 Open 1469501282 Open 1469501285 Open 1469501291 Open 1469501297 Open 1469501303 Open 1469501322 Open 1469501335 Open 1469501344 Open 1469501346 Open 1469501349 Open 1469501357 Open 1469501366 Open 1469501371 Open 1469501375 Open 1469501375 Open 1469501381 Open 1469501392 Open 1469501402 Open 1469501407 Open 1469501410 Open 1469501420 Open 1469501424 Open 1469501438 Close 1469501442 Open 1469501462 Open 1469501480 Open 1469501488 Open 1469501489 Open 1469501491 Open 1469501503 Open 1469501505 Open 1469501509 Open 1469501513 Open 1469501517 Open 1469501520 Open 1469501525 Open 1469501533 Open 1469501539 Open 1469501540 Open 1469501541 Open 1469501543 Open 1469501544 Close 1469501545 Open 1469501545 Open 1469501547 Open 1469501552 Open 1469501557 Open 1469501559 Open 1469501560 Open 1469501560 Open 1469501565 Open 1469501566 Open 1469501574 Open 1469501575 Open 1469501575 Open 1469501578 Open 1469501581 Open 1469501584 Open 1469501600 Open 1469501601 Open 1469501603 Open 1469501610 Open 1469501620 Open 1469501621 Open 1469501625 Close 1469501625 Open 1469501626 Open 1469501631 Open 1469501632 Open 1469501632 Open 1469501638 Open 1469501643 Open 1469501646 Open 1469501662 Open 1469501662 Open 1469501662 Open 1469501663 Open 1469501667 Open 1469501674 Open 1469501675 Close 1469501678 Open 1469501680 Open 1469501685 Open 1469501686 Open 1469501689 Open 1469501691 Open 1469501694 Close 1469501696 Open 1469501702 Open 1469501703 Open 1469501704 Open 1469501706 Open 1469501706 Open 1469501710 Open 1469501715 Open 1469501717 Open 1469501719 Open 1469501719 Open 1469501734 Open 1469501739 Open 1469501740 Open 1469501747 Open 1469501749 Close 1469501749 Open 1469501754 Open 1469501755 Open 1469501756 Open 1469501756 Open 1469501757 Open 1469501758 Open 1469501759 Open 1469501761 Open 1469501764 Open 1469501772 Open 1469501772 Close 1469501776 Open 1469501780 Open 1469501782 Open 1469501783 Open 1469501785 Open 1469501789 Open 1469501795 Open 1469501802 Open 1469501802 Open 1469501806 Open 1469501813 Open 1469501817 Open 1469501818 Close 1469501819 Open 1469501828 Open 1469501829 Open 1469501830 Open 1469501833 Open 1469501835 Open 1469501837 Open 1469501838 Open 1469501840 Open 1469501845 Open 1469501848 Open 1469501853 Open 1469501855 Close 1469501861 Open 1469501861 Open 1469501862 Open 1469501863 Open 1469501865 Open 1469501873 Open 1469501884 Open 1469501895 Open 1469501904 Open 1469501907 Close 1469501909 Open 1469501909 Open 1469501911 Open 1469501929 Open 1469501930 Open 1469501930 Open 1469501931 Open 1469501935 Open 1469501935 Open 1469501946 Open 1469501946 Open 1469501959 Open 1469501967 Close 1469501972 Open 1469501976 Open 1469501978 Open 1469501978 Open 1469501978 Open 1469501980 Open 1469501980 Open 1469501985 Open 1469501988 Open 1469501992 Open 1469501996 Open 1469502005 Open 1469502010 Close 1469502014 Open 1469502020 Open 1469502022 Open 1469502022 Open 1469502031 Open 1469502031 Open 1469502033 Open 1469502035 Open 1469502038 Open 1469502044 Open 1469502054 Open 1469502054 Open 1469502054 Open 1469502057 Open 1469502060 Open 1469502065 Open 1469502067 Open 1469502071 Open 1469502071 Close 1469502072 Open 1469502073 Open 1469502077 Open 1469502080 Open 1469502092 Open 1469502097 Open 1469502105 Open 1469502109 Open 1469502118 Open 1469502126 Open 1469502127 Open 1469502130 Open 1469502130 Open 1469502132 Open 1469502135 Open 1469502144 Open 1469502145 Open 1469502147 Close 1469502148 Open 1469502154 Open 1469502157 Open 1469502165 Open 1469502177 Open 1469502181 Open 1469502181 Open 1469502182 Open 1469502184 Open 1469502184 Open 1469502190 Open 1469502194 Open 1469502201 Open 1469502202 Open 1469502205 Open 1469502206 Open 1469502211 Open 1469502217 Open 1469502218 Open 1

In [0]:
display(spark.sql("SELECT * FROM IoT WHERE time < 1469501219"))

action time Open 1469501107 Open 1469501147 Open 1469501202